src/program/command_mode.rs

```diff
@@ -0,0 +1,300 @@
+    type Err = String;
+
+    /// All commands must implement FromStr
+    fn from_str(s: &str) -> Result<Commands, String> {
+        match s {
+            "sort" => Ok(Commands::Sort),
+            "ng" | "newglob" => Ok(Commands::NewGlob),
+            "h" | "help" => Ok(Commands::Help),
+            "q" | "quit" => Ok(Commands::Quit),
+            "r" | "reverse" => Ok(Commands::Reverse),
+            "df" | "destfolder" => Ok(Commands::DestFolder),
+            "m" | "max" => Ok(Commands::MaximumImages),
+            _ => Err(format!("Error: no such command \"{}\"", s)),
+        }
+    }
+}
+
+/// Converts the provided path by user to a path that can be glob'd
+/// Directories are changed from /home/etc to /home/etc/*
+fn convert_path_to_globable(path: &str) -> Result<String, String> {
+    let expanded_path =
+        full(path).map_err(|e| format!("Error: \"{}\": {}", e.var_name, e.cause))?;
+    let mut absolute_path = String::from(expanded_path);
+    // If path is a dir, add /* to glob
+    if PathBuf::from(&absolute_path).is_dir() {
+        if !absolute_path.ends_with('/') {
+            absolute_path.push('/');
+        }
+        absolute_path.push('*');
+    }
+    Ok(absolute_path)
+}
+
+/// Globs the passed path, returning an error if no images are in that path, glob::glob fails, or
+/// path is unexpected
+fn glob_path(path: &str) -> Result<Vec<PathBuf>, String> {
+    use crate::cli::push_image_path;
+
+    let mut new_images: Vec<PathBuf> = Vec::new();
+    let globable_path = convert_path_to_globable(path)?;
+    let path_matches = glob::glob(&globable_path).map_err(|e| e.to_string())?;
+    for path in path_matches {
+        match path {
+            Ok(p) => {
+                push_image_path(&mut new_images, p);
+            }
+            Err(e) => {
+                let err_msg = format!("Error: Unexpected path {}", e);
+                return Err(err_msg);
+            }
+        }
+    }
+    if new_images.is_empty() {
+        let err_msg = format!("Error: path \"{}\" had no images", path);
+        return Err(err_msg);
+    }
+    Ok(new_images)
+}
+
+/// Uses regex to parse user input, then concating all subsequent strings that are escaped together
+/// TODO: have regex capture the ending space in case of escaped files
+fn parse_user_input(input: &str) -> Vec<String> {
+    let mut input_vec: Vec<String> = Vec::new();
+    let re: Regex = Regex::new(r"[\x21-\x7E]+(\\)?").unwrap();
+    let mut escaped = false;
+
+    for regex_match in re.captures_iter(&input) {
+        let value: String = regex_match[0].to_string();
+        if value.ends_with('\\') && !escaped {
+            escaped = true;
+            input_vec.push(value);
+        } else if escaped {
+            if let Some(last) = input_vec.last_mut() {
+                last.push(' ');
+                last.push_str(&value);
+            }
+        } else {
+            input_vec.push(value);
+        }
+    }
+    input_vec
+}
```

Comment:
I'm not entirely clear on what the above function does. Is it for parsing user input, specifically for the path + glob, or something else? Why is it limited to ascii? You should probably mention that in the function name, that it is specific to paths. Also, guess this won't work on windows?

Reply:
Added better documentation, but you're completely right my code base isn't cross platform.

This functionality will come in the next push or so :heart:

Reply:
Fixed in latest push, completely OS agnostic

marking as resolved